SmartDeviceLink/SDLProtocol.m

```diff
@@ -87,15 +97,23 @@ - (UInt8)sdl_retrieveSessionIDforServiceType:(SDLServiceType)serviceType {
 #pragma mark - SDLTransportDelegate
 
 - (void)onTransportConnected {
-    for (id<SDLProtocolListener> listener in self.protocolDelegateTable.allObjects) {
+    NSArray<id<SDLProtocolListener>> *listeners;
+    @synchronized(self.protocolDelegateTable) {
+        listeners = self.protocolDelegateTable.allObjects;
+    }
+    for (id<SDLProtocolListener> listener in listeners) {
         if ([listener respondsToSelector:@selector(onProtocolOpened)]) {
             [listener onProtocolOpened];
         }
     }
 }
 
 - (void)onTransportDisconnected {
-    for (id<SDLProtocolListener> listener in self.protocolDelegateTable.allObjects) {
+    NSArray<id<SDLProtocolListener>> *listeners;
+    @synchronized(self.protocolDelegateTable) {
```

Comment:
Also unnecessary

Reply:
I don't know why, I cannot add any comment to your review comment above. So I'm adding here.

Well, Apple's document says that `NSHashTable` isn't a thread-safe class.
https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/Multithreading/ThreadSafetySummary/ThreadSafetySummary.html

I created a simple code that accesses to a `NSHashTable` instance from two different queues. I didn't see any issue if one of the queue is calling `addObject` while another is calling `allObjects`. However, if one of the queue is calling `addObject` and `removeObject`, and the other is calling `allObjects`, I saw the program crashes on the second queue.
My sample code is attached: [NSHashTableTest.zip](https://github.com/smartdevicelink/sdl_ios/files/2289745/NSHashTableTest.zip)

I do understand your concern that locking slows down the message processing. Another solution is to access `protocolDelegateTable` from a single queue, which is `SDLProtocol`'s `_receiveQueue`. When we call `addObject` or `removeObject` on the table, we do it inside `dispatch_sync(_receiveQueue, ^{ ... });`. Also, `onTransportConnected` and `onTransportDisconnected` will be updated to acquire the list of listeners inside `dispatch_sync(_receiveQueue, ^{ ... });`. In this way, I think we can achieve thread safety while removing the locks from the callbacks. I think this idea requires adding dedicated methods like `addProtocolListener` and `removeProtocolListener` in `SDLProtocol` class. So  I hesitated to implement it at the first place (since `SDLProtocol` was a public class at that time.)

If this sounds good for you, then I will investigate the code again and will update the code in such way.

Reply:
Thanks for doing that investigation, it's greatly appreciated. I agree that `@synchronized` is necessary unless we only access on one queue. Since we currently are not, (and I don't think we should heavily use `dispatch_sync` as that will probably be slower). I would also like to note, however, that we don't need to use `allObjects` to create an array. We can simply loop over the `table` as is (though it would have to be synchronized.

Reply:
That is true, however, the rest of the receive calls would need to be in a `dispatch_async`, which is likely to complicate our logic.